pkg/liqo-controller-manager/ip-controller/ip_controller.go

```diff
@@ -0,0 +1,203 @@
+	if err := r.Get(ctx, req.NamespacedName, &ip); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(4).Infof(" %q not found", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		klog.Errorf("an error occurred while getting IP %q: %v", req.NamespacedName, err)
+		return ctrl.Result{}, err
+	}
+
+	// Retrieve the remote cluster ID from the labels.
+	remoteclusterID := ip.Labels[consts.RemoteClusterID] // it should always be present thanks to validating webhook
+
+	_, err := foreignclusterutils.GetForeignClusterByID(ctx, r.Client, remoteclusterID)
+	if apierrors.IsNotFound(err) {
+		klog.Errorf("no ForeignCluster with clusterID %q found: %v", remoteclusterID, err)
+		return ctrl.Result{}, err
+	} else if err != nil {
+		klog.Errorf("an error occurred when trying to retrieve ForeignClusters: %v", err)
+		return ctrl.Result{}, err
+	}
+
+	desiredIP = ip.Spec.IP
+
+	if ip.GetDeletionTimestamp().IsZero() {
+		if !controllerutil.ContainsFinalizer(&ip, ipamIPFinalizer) {
+			// Check if IP is associated to an existing Network. Requirements:
+			// - Network must be associated to the same remote cluster
+			// - Network CIDR must contain the IP
+			// If a valid Network is found, it is set as ownerReference for the IP.
+
+			var nwList ipamv1alpha1.NetworkList
+			clusterIDSelector := client.MatchingLabelsSelector{
+				Selector: labels.Set{consts.RemoteClusterID: remoteclusterID}.AsSelector()}
+			if err := r.List(ctx, &nwList, clusterIDSelector); err != nil {
+				klog.Errorf("unable to list networks: %v", err)
+				return ctrl.Result{}, err
+			}
+
+			var ownerNetwork *ipamv1alpha1.Network
+			for i := range nwList.Items {
+				nw := &nwList.Items[i]
+				if networkContainsIP(nw.Spec.CIDR, desiredIP) {
+					ownerNetwork = nw
+					break
+				}
+			}
+			if ownerNetwork == nil {
+				err := fmt.Errorf("no network with clusterID %q contains IP %q (%s)", remoteclusterID, req.NamespacedName, desiredIP)
+				klog.Error(err)
+				return ctrl.Result{}, err
+			}
+
+			// Set found network as owner of the IP.
+			// The owner has the "foregroundDeletion" finalizer, so that all the IPs will be deleted before the Network.
+			if err := controllerutil.SetControllerReference(ownerNetwork, &ip, r.Scheme); err != nil {
+				klog.Errorf("error while setting owner reference to IP %q: %v", req.NamespacedName, err)
+				return ctrl.Result{}, err
+			}
+
+			// Add finalizer to prevent deletion without unmapping the IP.
+			controllerutil.AddFinalizer(&ip, ipamIPFinalizer)
+
+			// Update the IP object
+			if err := r.Update(ctx, &ip); err != nil {
+				klog.Errorf("error while adding finalizer to IP %q: %v", req.NamespacedName, err)
+				return ctrl.Result{}, err
+			}
+
+			klog.Infof("finalizer and network owner correctly added to IP %q", req.NamespacedName)
+
+			// We return immediately and wait for the next reconcile to eventually update the status.
+			return ctrl.Result{}, nil
+		}
+
+		// Update IP status if it is not set yet
```

Comment:
I would add a comment to specify that we do this check since the ipam logic is not idempotent